src/mem/alloc.h

```diff
@@ -747,35 +747,29 @@ namespace snmalloc
       {
         auto& bp = bump_ptrs[i];
         auto rsize = sizeclass_to_size(i);
-        FreeListHead ffl;
+        FreeListIter ffl;
         while (pointer_align_up(bp, SLAB_SIZE) != bp)
         {
           Slab::alloc_new_list(bp, ffl, rsize);
-          SlabNext* prev = ffl.value;
-          while (prev != nullptr)
+          while (!ffl.empty())
           {
-            auto n = Metaslab::follow_next(prev);
-            Superslab* super = Superslab::get(prev);
-            Slab* slab = Metaslab::get_slab(prev);
-            small_dealloc_offseted_inner(super, slab, prev, i);
-            prev = n;
+            auto curr = ffl.take();
+            Superslab* super = Superslab::get(curr);
+            Slab* slab = Metaslab::get_slab(curr);
+            small_dealloc_offseted_inner(super, slab, curr, i);
           }
         }
       }
 
       for (size_t i = 0; i < NUM_SMALL_CLASSES; i++)
       {
-        auto prev = small_fast_free_lists[i].value;
-        small_fast_free_lists[i].value = nullptr;
-        while (prev != nullptr)
+        while (!small_fast_free_lists[i].empty())
         {
-          auto n = Metaslab::follow_next(prev);
+          auto curr = small_fast_free_lists[i].take();
 
-          Superslab* super = Superslab::get(prev);
-          Slab* slab = Metaslab::get_slab(prev);
-          small_dealloc_offseted_inner(super, slab, prev, i);
-
-          prev = n;
+          Superslab* super = Superslab::get(curr);
```

Comment:
Not necessarily required for this PR, but looking ahead: it might be nice, while in the neighborhood, to somehow articulate that all objects in a `small_fast_free_list` are in the same `Slab`.  With the CHERI work, we're going to be storing allocation-bounded pointers in the `small_fast_free_list` and, because I was lazy here, we end up amplifying for each one to get its associated `Superslab` and `Slab`, but we should just need to do that once.  It's not a huge deal because this is just in the debugging path, tho'.

Reply:
So that is checked in the CHECK_CLIENT version for Integer pointers.  I think we can ensure `take` is doing some checking elsewhere?

Reply:
I'm content to leave it not explicitly checked on CHERI, since we'll blow up pretty fast if we take something out of bounds.  How'd you feel about code like
```
  if (!small_fast_free_lists[i].empty()) {
    auto first = small_fast_free_lists[i].peek(); // wrap with large_allocator.ptrauth_amplify() once AuthPtr<>s land
    auto slab = Metaslab::get_slab(first);
    auto super = Superslab::get(first);
    while (!small_fast_free_lists[i].empty()) {
      // ...
```

Reply:
Done